diff --git a/src/lib.rs b/src/lib.rs
index effc178..26c580d 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -52,7 +52,7 @@ mod macros;
 mod alloc;
 mod control;
 mod hasher;
-mod raw;
+pub mod raw;
 mod util;
 
 mod external_trait_impls;
diff --git a/src/raw_entry.rs b/src/raw_entry.rs
index 13b71bc..050c904 100644
--- a/src/raw_entry.rs
+++ b/src/raw_entry.rs
@@ -2,6 +2,7 @@ use crate::Equivalent;
 use crate::alloc::{Allocator, Global};
 use crate::map::{HashMap, equivalent, make_hash, make_hasher};
 use crate::raw::{Bucket, RawTable};
+use crate::set::HashSet;
 use core::fmt::{self, Debug};
 use core::hash::{BuildHasher, Hash};
 use core::mem;
@@ -159,6 +160,18 @@ impl<K, V, S, A: Allocator> HashMap<K, V, S, A> {
     }
 }
 
+impl<T, S, A: Allocator> HashSet<T, S, A> {
+    #[cfg_attr(feature = "inline-more", inline)]
+    pub fn raw_entry_mut(&mut self) -> RawEntryBuilderMut<'_, T, (), S, A> {
+        self.map.raw_entry_mut()
+    }
+
+    #[cfg_attr(feature = "inline-more", inline)]
+    pub fn raw_entry(&self) -> RawEntryBuilder<'_, T, (), S, A> {
+        self.map.raw_entry()
+    }
+}
+
 /// A builder for computing where in a [`HashMap`] a key-value pair would be stored.
 ///
 /// See the [`HashMap::raw_entry_mut`] docs for usage examples.
diff --git a/src/table.rs b/src/table.rs
index f50f575..5c486d9 100644
--- a/src/table.rs
+++ b/src/table.rs
@@ -49,7 +49,7 @@ pub struct HashTable<T, A = Global>
 where
     A: Allocator,
 {
-    pub(crate) raw: RawTable<T, A>,
+    pub raw: RawTable<T, A>,
 }
 
 impl<T> HashTable<T, Global> {
