diff --git a/src/lib.rs b/src/lib.rs
index 0ab5bdf..c673a08 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -155,6 +155,8 @@
 
 #![deny(warnings, missing_docs, missing_debug_implementations)]
 
+#![feature(const_convert, const_trait_impl, pattern_types, pattern_type_macro, structural_match)]
+
 //#![cfg_attr(not(feature = "std"), no_std)]
 #[cfg(not(feature = "std"))]
 compile_error!("`std` feature currently required, support for `no_std` may be added later");
diff --git a/src/request.rs b/src/request.rs
index 0eb36b7..66d1462 100644
--- a/src/request.rs
+++ b/src/request.rs
@@ -655,6 +655,16 @@ impl<T> Request<T> {
         (self.head, self.body)
     }
 
+    #[inline]
+    pub fn parts(&self) -> &Parts {
+        &self.head
+    }
+
+    #[inline]
+    pub fn parts_mut(&mut self) -> &mut Parts {
+        &mut self.head
+    }
+
     /// Consumes the request returning a new request with body mapped to the
     /// return type of the passed in function.
     ///
diff --git a/src/response.rs b/src/response.rs
index b8fc9b2..abdbd92 100644
--- a/src/response.rs
+++ b/src/response.rs
@@ -448,6 +448,16 @@ impl<T> Response<T> {
         (self.head, self.body)
     }
 
+    #[inline]
+    pub fn parts(&self) -> &Parts {
+        &self.head
+    }
+
+    #[inline]
+    pub fn parts_mut(&mut self) -> &mut Parts {
+        &mut self.head
+    }
+
     /// Consumes the response returning a new response with body mapped to the
     /// return type of the passed in function.
     ///
diff --git a/src/status.rs b/src/status.rs
index bc58ba4..233c348 100644
--- a/src/status.rs
+++ b/src/status.rs
@@ -17,7 +17,6 @@
 use std::convert::TryFrom;
 use std::error::Error;
 use std::fmt;
-use std::num::NonZeroU16;
 use std::str::FromStr;
 
 /// An HTTP status code (`status-code` in RFC 9110 et al.).
@@ -41,8 +40,41 @@ use std::str::FromStr;
 /// assert_eq!(StatusCode::NOT_FOUND.as_u16(), 404);
 /// assert!(StatusCode::OK.is_success());
 /// ```
-#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct StatusCode(NonZeroU16);
+#[derive(Clone, Copy)]//, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct StatusCode(pattern_type!(u16 is 100..=999));
+
+impl PartialEq for StatusCode {
+    #[inline]
+    fn eq(&self, other: &StatusCode) -> bool {
+        u16::from(*self) == u16::from(*other)
+    }
+}
+
+impl Eq for StatusCode {}
+
+impl PartialOrd for StatusCode {
+    #[inline]
+    fn partial_cmp(&self, other: &StatusCode) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for StatusCode {
+    #[inline]
+    fn cmp(&self, other: &StatusCode) -> core::cmp::Ordering {
+        u16::from(*self).cmp(&u16::from(*other))
+    }
+}
+
+impl core::hash::Hash for StatusCode {
+    #[inline]
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        u16::from(*self).hash(state)
+    }
+}
+
+impl core::marker::StructuralPartialEq for StatusCode {}
 
 /// A possible error value when converting a `StatusCode` from a `u16` or `&str`.
 ///
@@ -52,6 +84,20 @@ pub struct InvalidStatusCode {
     _priv: (),
 }
 
+impl const AsRef<u16> for StatusCode {
+    #[inline]
+    fn as_ref(&self) -> &u16 {
+        unsafe { core::ptr::from_ref(&self.0).cast::<u16>().as_ref_unchecked() }
+    }
+}
+
+impl const AsMut<u16> for StatusCode {
+    #[inline]
+    fn as_mut(&mut self) -> &mut u16 {
+        unsafe { core::ptr::from_mut(&mut self.0).cast::<u16>().as_mut_unchecked() }
+    }
+}
+
 impl StatusCode {
     /// Converts a u16 to a status code.
     ///
@@ -72,13 +118,16 @@ impl StatusCode {
     #[inline]
     pub const fn from_u16(src: u16) -> Result<StatusCode, InvalidStatusCode> {
         if let 100..=999 = src {
-            if let Some(code) = NonZeroU16::new(src) {
-                return Ok(StatusCode(code));
-            }
+            return Ok(unsafe { Self::from_u16_unchecked(src) });
         }
         Err(InvalidStatusCode::new())
     }
 
+    #[inline(always)]
+    pub const unsafe fn from_u16_unchecked(src: u16) -> StatusCode {
+        StatusCode(unsafe { core::mem::transmute(src) })
+    }
+
     /// Converts a `&[u8]` to a status code.
     pub fn from_bytes(src: &[u8]) -> Result<StatusCode, InvalidStatusCode> {
         if src.len() != 3 {
@@ -93,10 +142,7 @@ impl StatusCode {
             return Err(InvalidStatusCode::new());
         }
 
-        let status = (a * 100) + (b * 10) + c;
-        NonZeroU16::new(status)
-            .map(StatusCode)
-            .ok_or_else(InvalidStatusCode::new)
+        Ok(unsafe { Self::from_u16_unchecked((a * 100) + (b * 10) + c) })
     }
 
     /// Returns the `u16` corresponding to this `StatusCode`.
@@ -116,7 +162,7 @@ impl StatusCode {
     /// ```
     #[inline]
     pub const fn as_u16(&self) -> u16 {
-        self.0.get()
+        *self.as_ref()
     }
 
     /// Returns a &str representation of the `StatusCode`
@@ -132,7 +178,7 @@ impl StatusCode {
     /// ```
     #[inline]
     pub fn as_str(&self) -> &str {
-        let offset = (self.0.get() - 100) as usize;
+        let offset = (self.as_u16() - 100) as usize;
         let offset = offset * 3;
 
         // Invariant: self has checked range [100, 999] and CODE_DIGITS is
@@ -168,43 +214,43 @@ impl StatusCode {
     /// assert_eq!(status.canonical_reason(), Some("OK"));
     /// ```
     pub fn canonical_reason(&self) -> Option<&'static str> {
-        canonical_reason(self.0.get())
+        canonical_reason(self.as_u16())
     }
 
     /// Check if status is within 100-199.
     #[inline]
     pub fn is_informational(&self) -> bool {
-        (100..200).contains(&self.0.get())
+        (100..200).contains(self.as_ref())
     }
 
     /// Check if status is within 200-299.
     #[inline]
     pub fn is_success(&self) -> bool {
-        (200..300).contains(&self.0.get())
+        (200..300).contains(self.as_ref())
     }
 
     /// Check if status is within 300-399.
     #[inline]
     pub fn is_redirection(&self) -> bool {
-        (300..400).contains(&self.0.get())
+        (300..400).contains(self.as_ref())
     }
 
     /// Check if status is within 400-499.
     #[inline]
     pub fn is_client_error(&self) -> bool {
-        (400..500).contains(&self.0.get())
+        (400..500).contains(self.as_ref())
     }
 
     /// Check if status is within 500-599.
     #[inline]
     pub fn is_server_error(&self) -> bool {
-        (500..600).contains(&self.0.get())
+        (500..600).contains(self.as_ref())
     }
 }
 
 impl fmt::Debug for StatusCode {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt::Debug::fmt(&self.0, f)
+        fmt::Debug::fmt(self.as_ref(), f)
     }
 }
 
@@ -248,10 +294,10 @@ impl PartialEq<StatusCode> for u16 {
     }
 }
 
-impl From<StatusCode> for u16 {
+impl const From<StatusCode> for u16 {
     #[inline]
     fn from(status: StatusCode) -> u16 {
-        status.0.get()
+        unsafe { core::mem::transmute(status) }
     }
 }
 
@@ -307,7 +353,7 @@ macro_rules! status_codes {
         impl StatusCode {
         $(
             $(#[$docs])*
-            pub const $konst: StatusCode = StatusCode(unsafe { NonZeroU16::new_unchecked($num) });
+            pub const $konst: StatusCode = unsafe { StatusCode($num) };
         )+
 
         }
diff --git a/src/uri/path.rs b/src/uri/path.rs
index 8f9356e..c88c43c 100644
--- a/src/uri/path.rs
+++ b/src/uri/path.rs
@@ -377,6 +377,12 @@ impl hash::Hash for PathAndQuery {
     }
 }
 
+impl From<PathAndQuery> for (Bytes, u16) {
+    fn from(path_and_query: PathAndQuery) -> Self {
+        (path_and_query.data.into(), path_and_query.query)
+    }
+}
+
 // ===== PartialEq / PartialOrd =====
 
 impl PartialEq for PathAndQuery {
